mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala

```diff
@@ -329,8 +330,16 @@ private[spark] object GradientBoostedTrees extends Logging {
       //       However, the behavior should be reasonable, though not optimal.
       baseLearnerWeights(m) = learningRate
 
+      if (pre_predError.getStorageLevel != StorageLevel.NONE ){
```

Comment:
Got it, one thing I missed is that you're adding a call to `persist()` and that is the key. You've hardcoded `MEMORY_AND_DISK` though... you should probably match the persistence level of the input, if any, as other similar blocks of code do. This gives the caller a way to turn this off.

I don't see how this relates to running out of memory though ... if anything this causes you to run out of memory? well, it still seems like an improvement and that may be unrelated.

(Will `previousPredError` ever be persisted after the loop completes?)
(You can probably also dispense with checking the storage level is not NONE because `unpersist()` just does nothing in this case.)